Dockerfile

```diff
@@ -208,17 +208,42 @@ RUN --mount=type=cache,target=/root/.cache/go-build \
         GOBIN=/build/ GO111MODULE=on go install "github.com/tc-hib/go-winres@${GOWINRES_VERSION}" \
      && /build/go-winres --help
 
-FROM dev-base AS containerd
+# containerd
+FROM base AS containerd-src
+WORKDIR /usr/src/containerd
+RUN git init . && git remote add origin "https://github.com/containerd/containerd.git"
+# CONTAINERD_VERSION is used to build containerd binaries, and used for the
+# integration tests. The distributed docker .deb and .rpm packages depend on a
+# separate (containerd.io) package, which may be a different version as is
+# specified here. The containerd golang package is also pinned in vendor.mod.
+# When updating the binary version you may also need to update the vendor
+# version to pick up bug fixes or new APIs, however, usually the Go packages
+# are built from a commit from the master branch.
+ARG CONTAINERD_VERSION=v1.6.10
+RUN git fetch -q --depth 1 origin "${CONTAINERD_VERSION}" +refs/tags/*:refs/tags/* && git checkout -q FETCH_HEAD
+
+FROM base AS containerd-build
+WORKDIR /go/src/github.com/containerd/containerd
 ARG DEBIAN_FRONTEND
-RUN --mount=type=cache,sharing=locked,id=moby-containerd-aptlib,target=/var/lib/apt \
-    --mount=type=cache,sharing=locked,id=moby-containerd-aptcache,target=/var/cache/apt \
-        apt-get update && apt-get install -y --no-install-recommends \
-            libbtrfs-dev
-ARG CONTAINERD_VERSION
-COPY /hack/dockerfile/install/install.sh /hack/dockerfile/install/containerd.installer /
-RUN --mount=type=cache,target=/root/.cache/go-build \
-    --mount=type=cache,target=/go/pkg/mod \
-        PREFIX=/build /install.sh containerd
+ARG TARGETPLATFORM
+RUN xx-apt-get install -y --no-install-recommends gcc libbtrfs-dev libsecret-1-dev
+RUN --mount=from=containerd-src,src=/usr/src/containerd,rw \
+    --mount=type=cache,target=/root/.cache <<EOT
```

Comment:
For my understanding; is there any advantage of using `--mount` for the source code instead of using `FROM containerd-src` ? Both look to be from the same base-image. Only advantage would be that this stage would be smaller (not have the source files), but wondering if the mount actually could lead to cache busts 🤔

Reply:
It's more efficient to use mounts as it will build concurrently the `RUN xx-apt-get` step and `containerd-src` stage.

> but wondering if the mount actually could lead to cache busts 🤔

You mean `--mount=from=containerd-src`? If so no.

Reply:
I see we're using `/root/.cache` in multiple stages; some static, some not, and some with `CGO_ENABLE=1` and some without;

- I recall we had issues in the past with go and CGO and caching (leading to incorrect cache matches); will that be an issue? Does it need unique IDs for CGO on (or off)?
- I see we're changing the cache mount from `/root/.cache/go-build` to cache all of `/root/.cache`; what else will be in that cache location?

Reply:
> * I recall we had issues in the past with go and CGO and caching (leading to incorrect cache matches); will that be an issue? Does it need unique IDs for CGO on (or off)?

I think it's fine because go create an unique [action](https://github.com/golang/go/blob/master/src/cmd/go/internal/work/action.go) for each builder with a dedicated sub cache folder named like `go-buildxxxx`. I'm not quite sure about the fingerprint though but didn't encounter issues so far. We could also set a dedicated id to the mount:

```
--mount=type=cache,target=/root/.cache,id=containerd-$CGO_ENABLED
```

Reply:
Maybe @cpuguy83 recalls the issue. I think I may have run into it myself as well, where Go had a "false positive" for caching, and it led to pretty obscure failures, but it's been a while.

If it's indeed something to take into account when using cache for Go projects, this would be definitely something we should document in https://github.com/moby/buildkit/blob/v0.11.0-rc1/frontend/dockerfile/docs/reference.md#example-cache-go-packages

(also if users should not use `/root/.cache/go-build`, but `/root/.cache` instead, if Go indeed doesn't use `go-build` as a subdirectory)

Reply:
Yes, the go compiler can't determine if cached cgo needs to be recompiled if an "include" changes.

Especially problematic when cross compiling different architectures with the same cache.

Reply:
Yes indeed @cpuguy83, got the same issue with rust and have to set TARGETPLATFORM as cache id: https://github.com/crazy-max/rust-docker-cross/blob/2f6d5f93dedbf1b9c8d65a172df47796281841da/Dockerfile#L40